graph.go

```diff
@@ -0,0 +1,293 @@
+package dalec
+
+import (
+	"fmt"
+	"strings"
+	"sync"
+
+	"github.com/pkg/errors"
+	"github.com/pmengelbert/stack"
+	"golang.org/x/exp/constraints"
+	"k8s.io/apimachinery/pkg/util/sets"
+)
+
+type Graph struct {
+	target   string
+	specs    map[string]*Spec
+	ordered  orderedDeps
+	indices  map[string]int
+	vertices []*vertex
+	edges    sets.Set[dependency]
+}
+
+type dependency struct {
+	v    *vertex
+	w    *vertex
+	kind string
+}
+
+type cycle []*vertex
+type cycleList []cycle
+type orderedDeps []*Spec
+
+func (o orderedDeps) targetSlice(target ...string) []*Spec {
```

Comment:
can you name this function something more specific or perhaps add a comment to make it a little clearer what it does?

Reply:
done